### ISSUE TITLE
Add rule S6302: Policies granting all privileges are security-sensitive

### ISSUE THREAD

**Issue by pierre-loup-tristant-sonarsource** [SonarSource/SonarJS#3450]
# What

The goal of this rule is to detect JS/TS code using the AWS CDK that creates policies granting all privileges.
This rule has been previously implemented for Python. See [SONARPY-1153](https://sonarsource.atlassian.net/browse/SONARPY-1153).

# Detection logic

## [`PolicyStatement`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.PolicyStatement.html)

Policies are defined by classes such as `Policy` or `ManagedPolicy`. These classes accept a `PolicyDocument`, which contains multiple `PolicyStatement`s. This `PolicyStatement` class should be the main focus of this rule.

### Constructor

The [constructor](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.PolicyStatement.html) receives a single `PolicyStatementProps` parameters. The following checks should be applied to the parameters of the `PolicyStatementProps` object that the constructor receives:

Status | `actions` | `effect`
-- | -- | --
Noncompliant | <ul><li> `["*"]` (single element) </li><li>`["service:permission","*"]` (contains `*` amount other elements) | <ul><li>`iam.Effect.ALLOW`</li><li>Not set
Compliant | Anything other than <ul><li> `["*"]` (single element) </li><li>`["service:permission","*"]` (contains `*` amount other elements) | Any

**Example**
```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement({
    sid: "AllowActions",
    effect: iam.Effect.ALLOW, // secondary location
    actions: ["*"], // Noncompliant
    resources: ["arn:aws:iam:::user/*"]
})

new iam.PolicyStatement({
    sid: "AllowSomeActions",
    effect: iam.Effect.ALLOW
    actions: ["iam:GetAccountSummary"], // Compliant
    resources: ["arn:aws:iam:::user/*"]
})
```

### `fromJson`

Apart from the constructor, the static method `PolicyStatement.fromJson()` may also be used to create a `PolicyStatement`. It takes an object as first parameter, which should be checked as follows (using JSONPath syntax):

Status | `$.Action` | `$.Effect`
-- | -- | --
Noncompliant |  <ul><li>`*`</li><li>`["*"]` (single element)</li><li>`["service:permission","*"]` (contains `*` amount other elements) | <ul><li>`"Allow"` (case insensitive) </li><li>`iam.Effect.ALLOW`
Compliant |  Anythin other than <ul><li>`*`</li><li>`["*"]` (single element)</li><li>`["service:permission","*"]` (contains `*` amount other elements) | Any

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement.fromJson({
    "Sid": "AllowActions",
    "Effect": "Allow", // secondary location
    "Action": ["*"], // Noncompliant
    "Resource": "arn:aws:iam:::user/*"
}

new iam.PolicyStatement.fromJson({
    "Sid": "AllowSomeActions",
    "Effect": iam.Effect.ALLOW,
    "Action": ["iam:GetAccountSummary"], // Compliant
    "Resource": "arn:aws:iam:::user/*"
})
```

### `fromJson`

`PolicyDocument` also has a static method `PolicyDocument.fromJson()` which will also indirectly create `PolicyStatements`. This method also takes a object, where `$.Statement` contains a `PolicyStatement[]` which is equal to the input of `PolicyStatement.fromJson()`.

The checks applied to `PolicyStatement.fromJson()` must be applied to each element of `$.Statement`.

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

const policyDocument = {
    "Version": "2012-10-17",
    "Statement": [{
      "Sid": "AllowActionsList",
      "Effect": "Allow", // secondary location
      "Action": "*", // Noncompliant
      "Resource": ["arn:aws:iam:::user/*"]
    },
    {
      "Sid": "AllowActionsStar",
      "Effect": "Allow", // secondary location
      "Action": ["*"], // Noncompliant
      "Resource": ["arn:aws:iam:::user/*"]
    },
    {
      "Sid": "AllowSomeActions",
      "Effect": "Allow",
      "Action": ["iam:GetAccountSummary"], // Compliant
      "Resource": ["arn:aws:iam:::user/*"]
    }]
}

const policyDocumentFromJson = iam.PolicyDocument.fromJson(policyDocument)
```

## RSPEC

This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1313) contains the message and highlighting guidelines.
